### ISSUE TITLE
Vertical metrics not calculated correctly

### ISSUE THREAD

**Issue by aaronbell** [googlefonts/ufo2ft#347]
I'm doing test builds of [Cascadia Code](https://github.com/aaronbell/cascadia-code/tree/master/sources) from UFO and am not getting the correct winAscent / winDescent values (under ufo2ft 2.9.1). 

In the font head table, the yMax is listed as 2226 (thanks to the Abrevehook), however, the OS/2 winAscent is 1977. Similarly, the yMin is -750 and the OS/2 winDescent is 480. This is throwing up a 'FAIL' in the FontBakery testing. Weirdly, 1977 does not align with any glyph height in the font.

Reading through the code, it appears that the OS/2 winAscent, winDescent values are simply set by setting them to the hhea ascender and descender values. Should it not use the head yMin and yMax values instead? 

That also makes me wonder why the hhea values are not being calculated correctly either. Have yet to track that down. 

Any guidance on where I can look to track down the issue, or what might be causing this to happen?

**Comment by anthrotype** [comment]
The defaults are just that, when you don't like them you can always override by setting the appropriate UFO info attributes.
If anything -- just like everything else -- these should be better documented.
Even I don't remember the rules and had to look at the code, namely in ufo2ft/fontInfoData.py

https://github.com/googlefonts/ufo2ft/blob/4e6e7d8d9c78ba819307e18a9d5e9a60f3d5b0eb/Lib/ufo2ft/fontInfoData.py#L208-L240

https://github.com/googlefonts/ufo2ft/blob/4e6e7d8d9c78ba819307e18a9d5e9a60f3d5b0eb/Lib/ufo2ft/fontInfoData.py#L99-L110

***

Off-topic but I'm curious, why do you prefer to import fontmake.instantiator and ufo2ft from a custom python script instead of simply calling fontmake command line tool from a shell script or Makefile or similar? Is there something missing that you need?

**Comment by anthrotype** [comment]
So yes, the fallback values (i.e. if not explicitly set) for OS/2.usWinAscent and OS/2usWinDescent are set the same as the fallback values for hhea.ascender hhea.descender: i.e. respectively, info.ascender + typoLineGap, and info.descender.

> Should it not use the head yMin and yMax values instead?

I guess yes, if you care prefer to avoid any clipping in Windows GDI environment at the expense of unequal vertical metrics across different platforms.
Since this is a complicated matter, I prefer ufo2ft to be agnostic and leave it to the user to make their own assessments.

**Comment by madig** [comment]
I suppose the direct use is because I made it and Aaron is building on it :) Also, the build script is generating  several UFo variants directly, so it makes sense to have fine control over what happens I think.

**Comment by anthrotype** [comment]
Maybe we should split off fontmake.instantiator before it becomes even more widely used. I hope we made it clear that we don't make any promises as to the stability of that API.

**Comment by aaronbell** [comment]
Thanks. Yes, as @madig indicated, I am using it because I worked from his original code :). 

The reason all this came out is because I'm reviewing the font in FontBakery, which is throwing up warnings about the vertical metrics. It'd be nice if their review code aligned with what the other Google projects produce by default! Anyway, I'll close this for now.

**Comment by anthrotype** [comment]
> It'd be nice if their review code aligned with what the other Google projects produce by default!

agreed. Let's keep open, maybe someone will find the time to work on this

**Comment by anthrotype** [comment]
note to future self or willing soul: confirm that default vertical metrics as produced by ufo2ft produce warnings when running fonts through fontbakery and assess whether it makes sense to change the defaults accordingly
